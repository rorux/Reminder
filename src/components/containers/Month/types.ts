import { TRecord } from '@store/records/types';

export type TMonthProps = {
  year: number;
  month: number;
  recordList: Array<TRecord>;
};
