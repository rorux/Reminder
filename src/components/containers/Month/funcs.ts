import { DAY_MS } from '@utils/records/ComputedDays';

export const getCountDaysOnMonth = (year: number, month: number, firstDayThisMonth: Date) => {
  const firstDayNextMonth = month === 11 ? new Date(year + 1, 0, 1) : new Date(year, month + 1, 1);
  return Math.round(
    (Date.parse(String(firstDayNextMonth)) - Date.parse(String(firstDayThisMonth))) /
      1000 /
      3600 /
      24
  );
};

export const getDateParse = (date: Date, order: number, month: number | null = null) => {
  const dateParse = Date.parse(String(date)) + DAY_MS * order;
  if (month !== null) {
    const dateObj = new Date(dateParse);
    if (dateObj.getMonth() === month) return dateParse;
    else return null;
  } else return dateParse;
};

export const getTheFirstWeek = (first: Date) => {
  const firstWeekDay = first.getDay();
  const weekArray: Array<number | null> = [null, null, null, null, null, null, null];

  if (firstWeekDay === 0) weekArray[6] = getDateParse(first, 0);
  else {
    for (let i = firstWeekDay - 1, j = 0; i <= 6; i++, j++) {
      weekArray[i] = getDateParse(first, j);
    }
  }
  if (weekArray[6]) {
    const nextDay = new Date(weekArray[6] + DAY_MS);
    return { weekArray, nextDay };
  } else return { weekArray, nextDay: null };
};

export const getWeek = (first: Date | null) => {
  const thisMonth = first?.getMonth();
  const weekArray: Array<number | null> = [null, null, null, null, null, null, null];
  let nextDay: Date | null = null;

  if (first !== null) {
    let toDoEmpty = false;
    for (let i = 0; i <= 6; i++) {
      if (toDoEmpty) weekArray[i] = null;
      else {
        if (getDateParse(first, i, thisMonth) === null) {
          weekArray[i] = null;
          toDoEmpty = true;
        } else weekArray[i] = getDateParse(first, i, thisMonth);
      }
    }

    if (weekArray[6]) {
      const checkNextDay = new Date(weekArray[6] + DAY_MS);
      if (checkNextDay.getMonth() === thisMonth) nextDay = checkNextDay;
    }
  }

  return { weekArray, nextDay };
};
