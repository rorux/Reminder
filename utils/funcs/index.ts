export function getEnumKeyByEnumValue(myEnum: any, enumValue: number | string): string {
  const keys = Object.keys(myEnum).filter((x) => myEnum[x] == enumValue);
  return keys.length > 0 ? keys[0] : '';
}

export function getEnumValueByEnumKey(myEnum: any, enumKey: number | string): string {
  return myEnum[enumKey];
}
