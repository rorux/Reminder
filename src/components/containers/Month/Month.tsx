import React from 'react';
import { styled } from '@mui/material/styles';
import Stack from '@mui/material/Stack';
import Paper from '@mui/material/Paper';
import { grey } from '@mui/material/colors';
import { v1 as uuid } from 'uuid';
import { TMonthProps } from './types';
import { getTheFirstWeek, getWeek } from './funcs';
import { weekdays } from './constants';
import Week from '../Week';

const Item = styled(Paper)(({ theme }) => ({
  ...theme.typography.caption,
  padding: theme.spacing(1),
  textAlign: 'center',
  width: 32,
  height: 32,
}));

const Month: React.FC<TMonthProps> = ({ year, month, recordList }) => {
  const first = new Date(year, month, 1);

  const week1 = getTheFirstWeek(first);
  const week2 = getWeek(week1.nextDay);
  const week3 = getWeek(week2.nextDay);
  const week4 = getWeek(week3.nextDay);
  const week5 = getWeek(week4.nextDay);
  const week6 = getWeek(week5.nextDay);

  return (
    <div data-test="month">
      <Stack direction="row" sx={{ justifyContent: 'space-between' }}>
        {weekdays.map((weekday) => (
          <Item elevation={0} sx={{ color: grey[400] }} key={uuid()}>
            {weekday}
          </Item>
        ))}
      </Stack>
      <Week weekArray={week1.weekArray} recordList={recordList} />
      <Week weekArray={week2.weekArray} recordList={recordList} />
      <Week weekArray={week3.weekArray} recordList={recordList} />
      <Week weekArray={week4.weekArray} recordList={recordList} />
      <Week weekArray={week5.weekArray} recordList={recordList} />
      <Week weekArray={week6.weekArray} recordList={recordList} />
    </div>
  );
};

export default Month;
