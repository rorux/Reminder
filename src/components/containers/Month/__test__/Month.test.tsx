import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import Month from '../Month';
import { getCountDaysOnMonth, getTheFirstWeek, getDateParse, getWeek } from '../funcs';

Enzyme.configure({ adapter: new Adapter() });

describe('Testing Functions', () => {
  it('countDaysOnMonth', () => {
    expect(getCountDaysOnMonth(2022, 0, new Date(2022, 0, 1))).toBe(31);
    expect(getCountDaysOnMonth(2022, 1, new Date(2022, 1, 1))).toBe(28);
    expect(getCountDaysOnMonth(2022, 2, new Date(2022, 2, 1))).toBe(31);
    expect(getCountDaysOnMonth(2022, 3, new Date(2022, 3, 1))).toBe(30);
    expect(getCountDaysOnMonth(2022, 4, new Date(2022, 4, 1))).toBe(31);
    expect(getCountDaysOnMonth(2022, 5, new Date(2022, 5, 1))).toBe(30);
    expect(getCountDaysOnMonth(2022, 6, new Date(2022, 6, 1))).toBe(31);
    expect(getCountDaysOnMonth(2022, 7, new Date(2022, 7, 1))).toBe(31);
    expect(getCountDaysOnMonth(2022, 8, new Date(2022, 8, 1))).toBe(30);
    expect(getCountDaysOnMonth(2022, 9, new Date(2022, 9, 1))).toBe(31);
    expect(getCountDaysOnMonth(2022, 10, new Date(2022, 10, 1))).toBe(30);
    expect(getCountDaysOnMonth(2022, 11, new Date(2022, 11, 1))).toBe(31);
    expect(getCountDaysOnMonth(2023, 2, new Date(2023, 2, 1))).toBe(31);
  });

  it('getTheFirstWeek', () => {
    expect(getTheFirstWeek(new Date(2022, 0, 1))).toEqual({
      weekArray: [
        null,
        null,
        null,
        null,
        null,
        Date.parse(String(new Date(2022, 0, 1))),
        Date.parse(String(new Date(2022, 0, 2))),
      ],
      nextDay: new Date(2022, 0, 3),
    });
    expect(getTheFirstWeek(new Date(2022, 1, 1))).toEqual({
      weekArray: [
        null,
        Date.parse(String(new Date(2022, 1, 1))),
        Date.parse(String(new Date(2022, 1, 2))),
        Date.parse(String(new Date(2022, 1, 3))),
        Date.parse(String(new Date(2022, 1, 4))),
        Date.parse(String(new Date(2022, 1, 5))),
        Date.parse(String(new Date(2022, 1, 6))),
      ],
      nextDay: new Date(2022, 1, 7),
    });
    expect(getTheFirstWeek(new Date(2022, 3, 1))).toEqual({
      weekArray: [
        null,
        null,
        null,
        null,
        Date.parse(String(new Date(2022, 3, 1))),
        Date.parse(String(new Date(2022, 3, 2))),
        Date.parse(String(new Date(2022, 3, 3))),
      ],
      nextDay: new Date(2022, 3, 4),
    });
    expect(getTheFirstWeek(new Date(2022, 4, 1))).toEqual({
      weekArray: [null, null, null, null, null, null, Date.parse(String(new Date(2022, 4, 1)))],
      nextDay: new Date(2022, 4, 2),
    });
    expect(getTheFirstWeek(new Date(2022, 8, 1))).toEqual({
      weekArray: [
        null,
        null,
        null,
        Date.parse(String(new Date(2022, 8, 1))),
        Date.parse(String(new Date(2022, 8, 2))),
        Date.parse(String(new Date(2022, 8, 3))),
        Date.parse(String(new Date(2022, 8, 4))),
      ],
      nextDay: new Date(2022, 8, 5),
    });
  });

  it('getDateParse', () => {
    const day1 = 1643662800000;
    const day2 = day1 + 86400000;
    const day3 = day2 + 86400000;
    expect(getDateParse(new Date(2022, 1, 1), 0)).toBe(day1);
    expect(getDateParse(new Date(2022, 1, 1), 1)).toBe(day2);
    expect(getDateParse(new Date(2022, 1, 1), 2)).toBe(day3);

    const day20220227 = 1645909200000;
    const day20220228 = day20220227 + 86400000;

    expect(getDateParse(new Date(2022, 1, 27), 0, 1)).toBe(day20220227);
    expect(getDateParse(new Date(2022, 1, 27), 1, 1)).toBe(day20220228);
    expect(getDateParse(new Date(2022, 1, 27), 2, 1)).toBe(null);
  });

  it('getWeek', () => {
    expect(getWeek(null)).toEqual({
      weekArray: [null, null, null, null, null, null, null],
      nextDay: null,
    });

    expect(getWeek(new Date(2022, 1, 7))).toEqual({
      weekArray: [
        Date.parse(String(new Date(2022, 1, 7))),
        Date.parse(String(new Date(2022, 1, 8))),
        Date.parse(String(new Date(2022, 1, 9))),
        Date.parse(String(new Date(2022, 1, 10))),
        Date.parse(String(new Date(2022, 1, 11))),
        Date.parse(String(new Date(2022, 1, 12))),
        Date.parse(String(new Date(2022, 1, 13))),
      ],
      nextDay: new Date(2022, 1, 14),
    });

    expect(getWeek(new Date(2022, 1, 14))).toEqual({
      weekArray: [
        Date.parse(String(new Date(2022, 1, 14))),
        Date.parse(String(new Date(2022, 1, 15))),
        Date.parse(String(new Date(2022, 1, 16))),
        Date.parse(String(new Date(2022, 1, 17))),
        Date.parse(String(new Date(2022, 1, 18))),
        Date.parse(String(new Date(2022, 1, 19))),
        Date.parse(String(new Date(2022, 1, 20))),
      ],
      nextDay: new Date(2022, 1, 21),
    });

    expect(getWeek(new Date(2022, 1, 28))).toEqual({
      weekArray: [Date.parse(String(new Date(2022, 1, 28))), null, null, null, null, null, null],
      nextDay: null,
    });

    expect(getWeek(new Date(2022, 6, 25))).toEqual({
      weekArray: [
        Date.parse(String(new Date(2022, 6, 25))),
        Date.parse(String(new Date(2022, 6, 26))),
        Date.parse(String(new Date(2022, 6, 27))),
        Date.parse(String(new Date(2022, 6, 28))),
        Date.parse(String(new Date(2022, 6, 29))),
        Date.parse(String(new Date(2022, 6, 30))),
        Date.parse(String(new Date(2022, 6, 31))),
      ],
      nextDay: null,
    });

    expect(getWeek(new Date(2022, 7, 29))).toEqual({
      weekArray: [
        Date.parse(String(new Date(2022, 7, 29))),
        Date.parse(String(new Date(2022, 7, 30))),
        Date.parse(String(new Date(2022, 7, 31))),
        null,
        null,
        null,
        null,
      ],
      nextDay: null,
    });

    expect(getWeek(new Date(2022, 0, 31))).toEqual({
      weekArray: [Date.parse(String(new Date(2022, 0, 31))), null, null, null, null, null, null],
      nextDay: null,
    });
  });
});
