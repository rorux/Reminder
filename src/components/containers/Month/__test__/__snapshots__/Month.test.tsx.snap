// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component Should render Item component 1`] = `
<ContextProvider
  value={
    Object {
      "insert": [Function],
      "inserted": Object {
        "n8odea": true,
      },
      "key": "css",
      "nonce": undefined,
      "registered": Object {
        "css-n8odea": "font-family:\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif;font-weight:400;font-size:0.75rem;line-height:1.66;letter-spacing:0.03333em;padding:8px;text-align:center;width:32px;height:32px;",
      },
      "sheet": StyleSheet {
        "_insertTag": [Function],
        "before": null,
        "container": undefined,
        "ctr": 0,
        "insertionPoint": undefined,
        "isSpeedy": false,
        "key": "css",
        "nonce": undefined,
        "prepend": undefined,
        "tags": Array [],
      },
    }
  }
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-n8odea{font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:0.75rem;line-height:1.66;letter-spacing:0.03333em;padding:8px;text-align:center;width:32px;height:32px;}",
      }
    }
    data-emotion="css n8odea"
  />
  <ForwardRef(Paper)
    className="css-n8odea"
  />
</ContextProvider>
`;

exports[`Month component Should render Month component 1`] = `
<div
  data-test="month"
>
  <ForwardRef(Stack)
    direction="row"
    sx={
      Object {
        "justifyContent": "space-between",
      }
    }
  >
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81ac340-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Пн
    </Styled(ForwardRef(Paper))>
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81aea50-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Вт
    </Styled(ForwardRef(Paper))>
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81b1160-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Ср
    </Styled(ForwardRef(Paper))>
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81b1161-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Чт
    </Styled(ForwardRef(Paper))>
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81b1162-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Пт
    </Styled(ForwardRef(Paper))>
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81b1163-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Сб
    </Styled(ForwardRef(Paper))>
    <Styled(ForwardRef(Paper))
      elevation={0}
      key="b81b1164-7165-11ec-9a3d-9589ff4d878b"
      sx={
        Object {
          "color": "#bdbdbd",
        }
      }
    >
      Вс
    </Styled(ForwardRef(Paper))>
  </ForwardRef(Stack)>
  <Week
    recordList={
      Array [
        Object {
          "days": Array [
            1642366800000,
            1644872400000,
          ],
          "holidays": "afterHoliday",
          "id": "1",
          "monthday": "8",
          "period": "monthly",
          "title": "title of monthly remind",
        },
      ]
    }
    weekArray={
      Array [
        null,
        1643662800000,
        1643749200000,
        1643835600000,
        1643922000000,
        1644008400000,
        1644094800000,
      ]
    }
  />
  <Week
    recordList={
      Array [
        Object {
          "days": Array [
            1642366800000,
            1644872400000,
          ],
          "holidays": "afterHoliday",
          "id": "1",
          "monthday": "8",
          "period": "monthly",
          "title": "title of monthly remind",
        },
      ]
    }
    weekArray={
      Array [
        1644181200000,
        1644267600000,
        1644354000000,
        1644440400000,
        1644526800000,
        1644613200000,
        1644699600000,
      ]
    }
  />
  <Week
    recordList={
      Array [
        Object {
          "days": Array [
            1642366800000,
            1644872400000,
          ],
          "holidays": "afterHoliday",
          "id": "1",
          "monthday": "8",
          "period": "monthly",
          "title": "title of monthly remind",
        },
      ]
    }
    weekArray={
      Array [
        1644786000000,
        1644872400000,
        1644958800000,
        1645045200000,
        1645131600000,
        1645218000000,
        1645304400000,
      ]
    }
  />
  <Week
    recordList={
      Array [
        Object {
          "days": Array [
            1642366800000,
            1644872400000,
          ],
          "holidays": "afterHoliday",
          "id": "1",
          "monthday": "8",
          "period": "monthly",
          "title": "title of monthly remind",
        },
      ]
    }
    weekArray={
      Array [
        1645390800000,
        1645477200000,
        1645563600000,
        1645650000000,
        1645736400000,
        1645822800000,
        1645909200000,
      ]
    }
  />
  <Week
    recordList={
      Array [
        Object {
          "days": Array [
            1642366800000,
            1644872400000,
          ],
          "holidays": "afterHoliday",
          "id": "1",
          "monthday": "8",
          "period": "monthly",
          "title": "title of monthly remind",
        },
      ]
    }
    weekArray={
      Array [
        1645995600000,
        null,
        null,
        null,
        null,
        null,
        null,
      ]
    }
  />
  <Week
    recordList={
      Array [
        Object {
          "days": Array [
            1642366800000,
            1644872400000,
          ],
          "holidays": "afterHoliday",
          "id": "1",
          "monthday": "8",
          "period": "monthly",
          "title": "title of monthly remind",
        },
      ]
    }
    weekArray={
      Array [
        null,
        null,
        null,
        null,
        null,
        null,
        null,
      ]
    }
  />
</div>
`;
